Cubical/Basics/BinNat.agda

```diff
@@ -174,165 +44,330 @@ posInd {P} h1 hs ps = f ps
+-- transport a proof that is proved directly using refl for binary
+-- numbers to unary numbers. This is an example of program/data
+-- refinement; we can use univalence to prove properties about
+-- inefficient data-structures using efficient ones.
 
 -- Doubling structure
-record Double : Set (ℓ-suc ℓ-zero) where
+record Double {ℓ} (A : Set ℓ) : Set (ℓ-suc ℓ) where
   constructor dC
   field
-    carrier : Set
     -- doubling function computing 2 * x
-    double : carrier -> carrier
+    double : A → A
     -- element to double
-    elt : carrier
+    elt : A
 open Double
 
-private
+-- Compute: 2^n * x
+doubles : ∀ {ℓ} {A : Set ℓ} (D : Double A) → ℕ → A → A
+doubles D n x = iter n (double D) x
+
+doubleBinℕ : Binℕ → Binℕ
+doubleBinℕ binℕ0       = binℕ0
+doubleBinℕ (binℕpos x) = binℕpos (x0 x)
+
+Doubleℕ : Double ℕ
+Doubleℕ = dC doubleℕ n1024
+  where
   -- 1024 = 2^8 * 2^2 = 2^10
   n1024 : ℕ
   n1024 = doublesℕ 8 4
 
-  DoubleN : Double
-  DoubleN = dC ℕ doubleℕ n1024
-
+DoubleBinℕ : Double Binℕ
+DoubleBinℕ = dC doubleBinℕ bin1024
+  where
+  -- 1024 = 10000000000 = 2^10
   bin1024 : Binℕ
   bin1024 = binℕpos (x0 (x0 (x0 (x0 (x0 (x0 (x0 (x0 (x0 (x0 pos1))))))))))
 
-  DoubleBinN : Double
-  DoubleBinN = dC Binℕ doubleBinℕ bin1024
-
--- Compute: 2^n * x
-doubles : ∀ D → (n : ℕ) → carrier D → carrier D
-doubles D n x = iter n (double D) x
-
+-- As these don't commute strictly we have to prove it separately and
+-- insert it in the proof of DoubleBinℕ≡Doubleℕ below
+Binℕ→ℕdouble : (x : Binℕ) → doubleℕ (Binℕ→ℕ x) ≡ Binℕ→ℕ (doubleBinℕ x)
+Binℕ→ℕdouble binℕ0       = refl
+Binℕ→ℕdouble (binℕpos x) = refl
+
+-- We use the equality between Binℕ and ℕ to get an equality of
+-- doubling structures
+DoubleBinℕ≡Doubleℕ : PathP (λ i → Double (Binℕ≡ℕ i)) DoubleBinℕ Doubleℕ
+DoubleBinℕ≡Doubleℕ i =
+  dC (λ x → glue (λ { (i = i0) → doubleBinℕ x
+                    ; (i = i1) → doubleℕ x })
+                 (hcomp (λ j → λ { (i = i0) → Binℕ→ℕdouble x j
+                                 ; (i = i1) → doubleℕ x })
+                        (doubleℕ (unglue {φ = i ∨ ~ i} x))))
+     (transp (λ j → Binℕ≡ℕ (i ∨ ~ j)) i (Doubleℕ .elt))
+
+-- We can now use transport to prove a property that is too slow to
+-- check with unary numbers
 private
   -- 2^20 * e = 2^5 * (2^15 * e)
-
-  propDouble : Double → Set
+  propDouble : ∀ {ℓ} {A : Set ℓ} → Double A → Set ℓ
   propDouble D = doubles D 20 (elt D) ≡ doubles D 5 (doubles D 15 (elt D))
 
-  -- The property we want to prove that takes long to typecheck:
+  -- The property we want to prove that takes long to typecheck for ℕ:
   -- 2^10 * 1024 = 2^5 * (2^5 * 1024)
   --
-  -- propN : propDouble DoubleN
-  -- propN = refl
+  -- propDoubleℕ : propDouble Doubleℕ
+  -- propDoubleℕ = refl
 
   -- With binary numbers it is instant
-  propBin : propDouble DoubleBinN
-  propBin = refl
+  propDoubleBinℕ : propDouble DoubleBinℕ
+  propDoubleBinℕ = refl
+
+  -- Why is this not typechecking fast?
```

Comment:
Can some Agda expert help me understand why Agda so eagerly unfolds `propDouble Doubleℕ` in the below piece of code? Is there some way to locally block reduction of some definition without making it globally blocked? 

@Saizan @andreasabel

Reply:
@mortberg Agda does not take advantage of the fact that `propDouble` does not need to be reduced to typecheck this term.

Two terms are either syntactically equal or they get reduced to whnf and compared.
This is because in principle `propDouble foo` and `propDouble bar` could be equal even if `foo` and `bar` are not.

I have tried to set things up so that we could rely on the syntactic equality, but I end up with terms that internally have different representation of universe levels, unfortunately. Maybe syntactic equality should be a little more lenient here.

A simple way around reducing `propDouble` is to turn it into a record type with a single field though.

Reply:
Thanks! Turning it into a record worked! Can you explain why?

Reply:
Because record (or data) types do not reduce further, so they get compared by name and equality of arguments.

e.g. if `R` is a record type with one argument, then `R A` and `R B` are already in WHNF, and their equality is determined by comparing `A` and `B`.

Reply:
Excellent, thanks! So whenever I have computationally expensive term that shouldn't be reduced I'll wrap it in a record (in Coq we do something similar using Modules).

Reply:
FYI you can also put functions under an `abstract` block to prevent them from being reduced.

Reply:
Yes, but that makes them globally blocked without the possibility of being un-blocked again right?

Reply:
IIUC yes.